File: erlang/seminars/transport/src/hub.erl
-module(hub).

-export([start/0]).

-include("frame.hrl").

start() ->
    {ok, spawn(fun() -> init() end)}.

init() ->
    hub([]).

hub(Connected) ->
    receive 
	{connect, Pid} ->
	    Ref = erlang:monitor(process, Pid),
	    hub([{Ref, Pid}|Connected]);

	{disconnect, Pid} ->
	    erlang:demonitor(process, Pid),
	    hub(lists:keydelete(Pid, 2, Connected));

	{'DOWN', Ref, process, _, _}  ->
	    hub(lists:keydelete(Ref, 1, Connected));
	
	#frame{}=Frm ->
	    lists:foreach(fun({_,Pid}) -> Pid ! Frm end, Connected),
	    hub(Connected);
	
	status ->
	    io:format("connect to ~w~n", [Connected]),
	    hub(Connected);	    

	quit ->
	    ok

	end.

File: exercises/transport/src/hub.ex
defmodule Hub do

  require Link
  
  def start() do
    {:ok, spawn(fn() -> init() end)}
  end

  def init() do
    :io.format("hub ~w: started~n", [self()])
    hub([])
  end

  def hub(connected) do
    receive do
      {:connect, pid} ->
	:io.format("hub ~w: connecting to ~w~n", [self(), pid])
	ref = :erlang.monitor(:process, pid)
	hub([{ref, pid}|connected])

      {:disconnect, pid} ->
	:io.format("hub ~w: disconnect ~w~n", [self(), pid])
	:erlang.demonitor(:process, pid)
	hub(List.keydelete(connected, pid, 1))

      {:DOWN, ref, :process, _, _}  ->
	:io.format("hub ~w: died ~w~n", [self(), ref])
	hub(List.keydelete(connected, ref, 0))

      Link.frame() = frm ->
	Enum.each(connected, fn({_,pid}) -> send(pid, frm) end)
	hub(connected)
	
      :status ->
	:io.format("hub ~w: connected to ~w~n", [self(), connected])
	hub(connected)

      :quit ->
	:ok

    end
  end

end
 
